### PR TITLE
Add more multi threading supports

### PR DESCRIPTION
**Pull request by xukai92**
This PR adds multi-threading for GP, GE, PIPE and MC as it is done for the cross entropy method in https://github.com/sisl/ExprOptimization.jl/pull/23.

The change for MC is more involved and probably needs some review.

### CODE REVIEW

File: src/MonteCarlo/MonteCarlos.jl [hunk 1]
```diff
@@ -73,16 +73,26 @@ Expression tree optimization using Monte Carlo with parameters p, grammar 'gramm
 function monte_carlo(p::MonteCarlo, grammar::Grammar, typ::Symbol, loss::Function; 
     verbose::Bool=false)
     dmap = mindepth_map(grammar)
-    best_tree, best_loss = RuleNode(0), Inf
```
Comment (xukai92) [inline src/MonteCarlo/MonteCarlos.jl:76]: The original implementation is memory-efficient and friendly for `_update_tracker!`, which cannot be assumed to be thread-safe.

File: src/MonteCarlo/MonteCarlos.jl [hunk 2]
```diff
@@ -73,16 +73,26 @@ Expression tree optimization using Monte Carlo with parameters p, grammar 'gramm
...
-    for i = 1:p.num_samples
-        verbose && println("samples: $i of $(p.num_samples)")
+
+    trees = Vector{RuleNode}(undef, p.num_samples)
```
Comment (xukai92) [inline src/MonteCarlo/MonteCarlos.jl:77]: I could have only allocate a vector of length `Threads.nthreads()`.
It would not be a problem for this loop but is an issue for the second loop for the `_update_tracker!` function, which would need to loops over all individual tree and loss.

Comment (rcnlee) [reply src/MonteCarlo/MonteCarlos.jl:77]: Yep, I agree with the changes that you've made.  It looks good to me.  Thanks for contributing!  I'll merge it.
